stylo: Bug 1374233 - Clamp interpolated values for properties which need to be restricted

BorisChiou opened the pull request:
Some properties only accept non-negative values, or values greater than or equal to one. It is possible to produce an negative interpolated values while using negative timing functions, so we have to apply a restriction to these values to avoid getting invalid values.

For example, line-height must be non-negative, but the output progress of some timing functions (e,g. cubic-bezier(0.25, -2, 0.75, 1)) may be a negative value, so the interpolated result of line-height is also negative.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix Bug 1374233.
- [X] These changes do not require tests because we have tests in Gecko side already.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17783)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/inherited_text.mako.rs, components/style/values/computed/text.rs, components/style/properties/longhand/padding.mako.rs, components/style/properties/longhand/position.mako.rs, components/style/properties/data.py and 13 more
 * @canaltinova: components/style/properties/longhand/inherited_text.mako.rs, components/style/values/computed/text.rs, components/style/properties/longhand/padding.mako.rs, components/style/properties/longhand/position.mako.rs, components/style/properties/data.py and 13 more
 * @emilio: components/style/properties/longhand/inherited_text.mako.rs, components/style/values/computed/text.rs, components/style/properties/longhand/padding.mako.rs, components/style/properties/longhand/position.mako.rs, components/style/properties/data.py and 13 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

BorisChiou commented:
r? nox

BorisChiou commented:
r? @nox

nox commented:
So my first issue with this is: even if we clamp at conversion-time, what about the result of interpolation? When does *that* get clamped?

BorisChiou commented:
cc @hiikezoe

nox commented on components/style/properties/longhand/font.mako.rs:
```diff
@@ -1081,6 +1084,16 @@ ${helpers.single_keyword_system("font-variant-caps",
+
+        impl AnimatedValueAsComputed for T {
+            #[inline]
+            fn restrict_value(self, restriction_type: Restriction) -> Self {
```
AFAIK there is never a need for such a method given Animatable methods can clamp directly.

nox commented on components/style/values/animated/mod.rs:
```diff
@@ -9,15 +9,35 @@
+/// Restriction types.
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Restriction {
```
We already have a type for that, please replace by `AllowedNumericType`.

nox commented on components/style/values/generics/effects.rs:
```diff
@@ -60,12 +61,87 @@ pub enum Filter<Angle, Factor, Length, DropShadow> {
+    }}
+}
+
+impl<Angle, Factor, Length, DropShadow> ToAnimatedValue for Filter<Angle, Factor, Length, DropShadow>
```
No, define the `Factor` impl yourself instead of clamping from here.

nox commented on components/style/values/generics/effects.rs:
```diff
@@ -77,6 +153,37 @@ pub struct SimpleShadow<Color, SizeLength, ShapeLength> {
+            color: ToAnimatedValue::from_animated_value(animated.color),
+            horizontal: ToAnimatedValue::from_animated_value(animated.horizontal),
+            vertical: ToAnimatedValue::from_animated_value(animated.vertical),
+            blur: convert_and_clamp_nonnegative_animated_value!(animated.blur)
```
Make the blur value its own wrapper type that clamps correctly.

nox commented on components/style/values/computed/text.rs:
```diff
@@ -61,3 +62,18 @@ impl Animatable for LineHeight {
     }
 }
+
+impl AnimatedValueAsComputed for LineHeight {
```
Why doesn't the computed type clamp itself during animation?

BorisChiou commented:
We don't change the result of interpolation. All we need to do is make sure we put the correct non-negative result into ServoComputedValues (by uncompute() function), with Animation/Transition cascade level, so we can see the correct animation on the screen.

The out of range interpolated values are still needed to be preserved during calculating AnimationValue in the case for additive or accumulative animations.

nox commented on components/style/values/computed/background.rs:
```diff
@@ -50,3 +51,16 @@ impl Animatable for BackgroundSize {
     }
 }
+
+impl AnimatedValueAsComputed for BackgroundSize {
```
Why doesn't the computed type clamp itself during animation?

nox commented on components/style/values/animated/mod.rs:
```diff
@@ -9,15 +9,35 @@
 use app_units::Au;
+use std::cmp;
 use values::computed::Angle as ComputedAngle;
+use values::computed::BorderCornerRadius as ComputedBorderCornerRadius;
```
Please no, make a submodule `border` just like there is one in `computed` and `specified`.

emilio replied:
This still isn't addressed, I think.

nox commented:
> We don't change the result of interpolation. All we need to do is make sure we put the correct non-negative result into ServoComputedValues (by uncompute() function), with Animation/Transition cascade level, so we can see the correct animation on the screen.

Why wasn't the computed value clamped itself, then?

BorisChiou commented:
Per our talk on irc, I will try to add new types for those non-negative or greater-than-or-equal-to-one, to avoid additional conditional clamping.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17774) made this pull request unmergeable. Please resolve the merge conflicts.

BorisChiou committed:
Bug 1374233 - Part 1: Add NonNegativeNumber and GreaterThanOrEqualToOneNumber.

NonNegativeNumber: for -moz-box-flex, flex-grow, and flex-shrink.
GreaterThanOrEqualToOneNumber: for stroke-miterlimit.

MozReview-Commit-ID: Kgbt99BPdVA

BorisChiou committed:
Bug 1374233 - Part 2: Add NonNegativeAu.

Add values::computed::NonNegativeAu, so BorderSideWith could be computed
to this non-negative Au, for the following properties:
1. outline-width
2. border-{*}-width
3. column-rule-width
4. -webkit-text-stroke-width

MozReview-Commit-ID: ASHaB2F7VtM

BorisChiou committed:
Bug 1374233 - Part 4: Add NonNegativeLengthOr{Auto|Normal|Number}.

Add NonNegativeLength, which could be computed to NonNegativeAu. So we
can declare Either<NonNegativeLength, X>, X=Auto, Normal, or Number.

NonNegativeLengthOrAuto is for column-width.
NonNegativeLengthOrNormal is for column-gap.
NonNegativeLengthOrNumber is for -moz-tab-size.

MozReview-Commit-ID: AfU8XpA1um0

BorisChiou committed:
Bug 1374233 - Part 6: Add PositiveInteger and PositiveIntegerOrAuto for column-count.

column-count should be a positive integer or auto.

MozReview-Commit-ID: 9LFvrYo8De5

BorisChiou committed:
Bug 1374233 - Part 7: Add NonNegativeLengthOrPercentage.

For padding-{*} and grid-{*}-gap.

MozReview-Commit-ID: 81G0b1k6JnD

BorisChiou committed:
Bug 1374233 - Part 8: Implement ToAnimatedValue for BorderCornerRadius.

BorderCornerRadius should always be non-negative, so we can implemennt
ToAnimatedValue for it directly, for properties:
1. border-{*}-radius
2. -moz-outline-{*}-radius

MozReview-Commit-ID: HEbeHz9Hfkd

BorisChiou committed:
Bug 1374233 - Part 5: Use NonNegativeLength and NonNegativeAu for border-spacing.

We already have NonNegativeLength and NonNegativeAu, so we can re-use it
to define the specified value and the computed value of border-spacing.
And then implement ToAnimatedValue for it.

MozReview-Commit-ID: CLckpKMYVXU

BorisChiou committed:
Bug 1374233 - Part 3: Use NonNegativeAu as computed values for font-size related properties.

For font-size and font-size-adjust.

MozReview-Commit-ID: 5rrfVSzB7WF

BorisChiou committed:
Bug 1374233 - Part 9: Add NonNegativeLengthOrPercentageOrNumber for stroke-{width|dasharry}.

Also add SVGWidth which use NonNegativeSVGLengthOrPercentage
as the LengthType for stroke-width.

MozReview-Commit-ID: 8gD0fVe2eAe

BorisChiou committed:
Bug 1374233 - Part 10: Implement ToAnimatedValue for MozLength and MaxLength.

For flex-basis, width/height, {max|min}-width, {max|min}-height.

MozReview-Commit-ID: 4gGYSXoBS8e

BorisChiou committed:
Bug 1374233 - Part 11: Implement ToAnimatedValue for background-size.

MozReview-Commit-ID: DMcvpaqHdy9

BorisChiou committed:
Bug 1374233 - Part 12: Implement ToAnimatedValue for LineHeight.

Besides, we replace its type with
GenericLineHeight<NonNegativeNumber, NonNegativeAu>.

MozReview-Commit-ID: GGOGXyUFJsJ

BorisChiou committed:
Bug 1374233 - Part 13: Use NonNegative{*} types for components of Shadow and Filter.

MozReview-Commit-ID: Im4KGy1n9IJ

BorisChiou commented:
I just rewrote my patches locally (by introducing some non-negative types and greater-than-or-equal-to-one types, and updating some existing types, e.g. LineHeight), so will update this PR soon.

BorisChiou commented:
Gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=d9faa25293b43b3fc8ec44cf2e7f0f1194e10f6e

BorisChiou commented:
r? @nox again. Thanks.
I add some types for NonNegtaive and Greater-than-Or-Equal-To-One values, and then use these types as the specified/computed values of these properties which need to be clamped.

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e066336b1570671d42b730da6a5b9f3acf526a67 with merge 6a757ddc9e84ca28297c2783479d3d3c97190c1f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8068), [arm32](http://build.servo.org/builders/arm32/builds/8291), [arm64](http://build.servo.org/builders/arm64/builds/8324), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8410), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5171), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5205), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8405), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/882), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/881), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5107), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5165), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/883), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/888), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3649)
State: approved= try=True

BorisChiou commented:
There are some conflicts now, but I'd like to rebase my patches after making sure the design is accepted or knowing what should I fix.

BorisChiou commented:
@emilio, nox is still on PTO, so do you feel comfortable with reviewing these patches?

emilio commented on components/style/properties/gecko.mako.rs:
```diff
@@ -392,6 +392,20 @@ impl ${style_struct.gecko_struct_name} {
+<%def name="impl_simple_wrapper_setter(ident, gecko_ffi_name)">
+    #[allow(non_snake_case)]
+    pub fn set_${ident}(&mut self, v: longhands::${ident}::computed_value::T) {
+        ${set_gecko_property(gecko_ffi_name, "v.0")}
```
We have something to implement stuff with conversions. Perhaps we should implement `From` and remove this mako code instead, wdyt?

BorisChiou replied:
Sounds great, I can try it.

emilio commented on components/style/values/specified/mod.rs:
```diff
@@ -507,6 +507,68 @@ impl ToCss for Number {
+
+    #[inline]
+    fn from_computed_value(computed: &Self::ComputedValue) -> Self {
+        GreaterThanOrEqualToOneNumber(Number::from_computed_value(&computed.0))
```
Ick, that's a decent amount of boilerplate...

emilio commented on components/style/properties/helpers/animated_properties.mako.rs:
```diff
@@ -907,6 +908,32 @@ impl Animatable for i32 {
+impl Animatable for NonNegativeNumber {
+    #[inline]
+    fn add_weighted(&self, other: &Self, self_portion: f64, other_portion: f64) -> Result<Self, ()> {
+        self.0.add_weighted(&other.0, self_portion, other_portion).map(NonNegativeNumber)
```
So, we're not clamping here, and I think that's the intention, right? If so, some comments about it would be nice.

BorisChiou replied:
Yes, it's the intention. I will add some comment here. Thanks

emilio commented on components/style/values/specified/mod.rs:
```diff
@@ -507,6 +507,68 @@ impl ToCss for Number {
+no_viewport_percentage!(GreaterThanOrEqualToOneNumber);
+
+impl Parse for GreaterThanOrEqualToOneNumber {
+    fn parse<'i, 't>(context: &ParserContext, input: &mut Parser<'i, 't>) -> Result<Self, ParseError<'i>> {
```
Can we remove `Number::parse_at_least_one` and `Number::parse_non_negative` now?

BorisChiou replied:
Not all non-negative properties are animated, so I didn't replace the type of all non-negative properties with NonNegativeNumber. I will try to check it.

BorisChiou replied:
OK. I can do it, but if you don't mind, I'd like to remove Number::parse_at_least_one() and Number::parse_non_negative() in a follow-up PR to keep this not too complicated.

emilio commented on components/style/properties/longhand/inherited_table.mako.rs:
```diff
@@ -66,20 +66,37 @@ ${helpers.single_keyword("caption-side", "top bottom",
+
+            #[inline]
+            fn from_animated_value(animated: Self::AnimatedValue) -> Self {
+                T { horizontal: NonNegativeAu::from_animated_value(animated.horizontal),
```
nit: indentation here is weird.

```
T {
    horizontal: ...,
    vertical: ...,
}
```

emilio reviewed (commented):
Looks fine over all. There's a fair amount of boilerplate (disclaimer: I don't know how previous versions of the patch were). Most of it could go away with `derive`, I think.

I wonder if some kind of `NonNegative<T>`, etc would be doable.

This also needs a rebase.

@nox does this patch address your comments properly? I'm not aware of the full discussion. I know you're on PTO, but it'd be nice if you could take a quick look. Thanks!

BorisChiou commented:
```NonNegative<T>``` might be a good way to refactor these patches (in the same PR or others). I'd like to know what nox think.

BorisChiou commented:
I just tried to use ```NonNegative<T>``` as the type of NonNegativeXXX, and it looks much better because lots of code could be written as template. I'd like to update a newer version by this way. Thanks for the suggestion, Emilio.

BorisChiou commented:
gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=a2db4c2a8d83bd0d837350505cb5046edb029f0c

The test failures on the try is not related to these patches, so the overall status looks good now.

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 8651acd94cd068d5c2f340ab447926550a3238d0 with merge d1f22c09581c8de7c4ef1d3519e2d9fee18dbe05...

BorisChiou commented:
I rebased the patches and use ```NonNegative<T>``` as the boilerplate to remove lots of redundant code.

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8187), [arm32](http://build.servo.org/builders/arm32/builds/8410), [arm64](http://build.servo.org/builders/arm64/builds/8443), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8529), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5292), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5325), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8526), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1002), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1005), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5227), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5285), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1004), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1008), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3768)
State: approved= try=True

nox commented:
r=me if geckotry is green.

BorisChiou commented:
Thanks for the review. I just push these commits to gecko try again. (Gecko try tree is closed now, so I will put the try link later.)

BorisChiou commented:
Gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=297b4aab13b3

BorisChiou commented:
@bors-servo r=nox

bors-servo commented:
:pushpin: Commit 8651acd has been approved by `nox`

<!-- @bors-servo r=nox 8651acd94cd068d5c2f340ab447926550a3238d0 -->

BorisChiou commented:
@bors-servo p=1

bors-servo commented:
:hourglass: Testing commit 8651acd94cd068d5c2f340ab447926550a3238d0 with merge 016ea11cbaf7586db144be122581c5b110649aee...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5356)

jdm commented:
That's #17168; you can retry.

BorisChiou commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8217), [arm32](http://build.servo.org/builders/arm32/builds/8441), [arm64](http://build.servo.org/builders/arm64/builds/8474), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8560), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5323), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8557), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1033), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5257), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5315), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1039), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3799) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5356), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1036), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1035)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8217), [arm32](http://build.servo.org/builders/arm32/builds/8441), [arm64](http://build.servo.org/builders/arm64/builds/8474), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8560), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5323), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5357), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8557), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1033), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1036), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5257), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5315), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1035), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1039), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3799)
Approved by: nox
Pushing 016ea11cbaf7586db144be122581c5b110649aee to master...
